File: packages/sdk-go/webhook_signature_test.go
package driftstack

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"strconv"
	"strings"
	"testing"
	"time"
)

func sign(body []byte, secret string, ts int64) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(strconv.FormatInt(ts, 10)))
	mac.Write([]byte("."))
	mac.Write(body)
	return "t=" + strconv.FormatInt(ts, 10) + ",v1=" + hex.EncodeToString(mac.Sum(nil))
}

func sigHex(body []byte, secret string, ts int64) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(strconv.FormatInt(ts, 10)))
	mac.Write([]byte("."))
	mac.Write(body)
	return hex.EncodeToString(mac.Sum(nil))
}

// During a secret-rotation grace window the server dual-signs into ONE
// header (t=,v1=<new>,v1=<old>); a verifier holding EITHER secret must pass.
func TestVerifyWebhookSignature_DualV1Rotation(t *testing.T) {
	body := []byte(`{"event":"x"}`)
	now := time.Now()
	ts := now.Unix()
	const newSecret = "whsec_new_rotated"
	const oldSecret = "whsec_old_pre_rotation"
	header := "t=" + strconv.FormatInt(ts, 10) + ",v1=" + sigHex(body, newSecret, ts) + ",v1=" + sigHex(body, oldSecret, ts)
	if !VerifyWebhookSignature(body, header, newSecret, VerifyWebhookOptions{Now: now}) {
		t.Error("new-secret holder should verify (first v1=, previously discarded)")
	}
	if !VerifyWebhookSignature(body, header, oldSecret, VerifyWebhookOptions{Now: now}) {
		t.Error("old-secret holder should verify (last v1=)")
	}
	if VerifyWebhookSignature(body, header, "whsec_unrelated", VerifyWebhookOptions{Now: now}) {
		t.Error("unrelated secret must not verify against any v1=")
	}
}

func TestVerifyWebhookSignature_RoundTrip(t *testing.T) {
	t.Parallel()
	body := []byte(`{"id":"evt-1","type":"session.completed","data":{}}`)
	secret := "whsec_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
	now := time.Now()
	header := sign(body, secret, now.Unix())
	if !VerifyWebhookSignature(body, header, secret, VerifyWebhookOptions{Now: now}) {
		t.Fatal("expected valid signature to verify")
	}
}

func TestVerifyWebhookSignature_TamperedBody(t *testing.T) {
	t.Parallel()
	body := []byte(`{"id":"evt-1"}`)
	secret := "whsec_xx"
	now := time.Now()
	header := sign(body, secret, now.Unix())
	if VerifyWebhookSignature([]byte(`{"id":"tampered"}`), header, secret, VerifyWebhookOptions{Now: now}) {
		t.Fatal("expected tampered body to fail verification")
	}
}

func TestVerifyWebhookSignature_WrongSecret(t *testing.T) {
	t.Parallel()
	body := []byte("hello")
	now := time.Now()
	header := sign(body, "whsec_correct", now.Unix())
	if VerifyWebhookSignature(body, header, "whsec_wrong", VerifyWebhookOptions{Now: now}) {
		t.Fatal("expected wrong secret to fail")
	}
}

func TestVerifyWebhookSignature_OldTimestamp(t *testing.T) {
	t.Parallel()
	body := []byte("x")
	secret := "whsec_xx"
	now := time.Now()
	tenMinAgo := now.Add(-10 * time.Minute)
	header := sign(body, secret, tenMinAgo.Unix())
	if VerifyWebhookSignature(body, header, secret, VerifyWebhookOptions{Now: now}) {
		t.Fatal("expected old timestamp to fail (default 5min tolerance)")
	}
}

func TestVerifyWebhookSignature_FutureTimestamp(t *testing.T) {
	t.Parallel()
	body := []byte("x")
	secret := "whsec_xx"
	now := time.Now()
	farFuture := now.Add(10 * time.Minute)
	header := sign(body, secret, farFuture.Unix())
	if VerifyWebhookSignature(body, header, secret, VerifyWebhookOptions{Now: now}) {
		t.Fatal("expected future timestamp to fail")
	}
}

// The skew comparison is exclusive, so the boundary itself is inside the window. The
// two arms above sit ten minutes out against a five-minute tolerance — nowhere near
// the edge — so an operator flip or a widened window passes both. These bracket it on
// each side. sdk-typescript and sdk-python carry the same points, because a customer
// picks one SDK and gets whatever that one enforces.
func TestVerifyWebhookSignature_AtToleranceEdge(t *testing.T) {
	t.Parallel()
	body := []byte("edge")
	secret := "whsec_xx"
	now := time.Now().Truncate(time.Second)
	atEdge := now.Add(-DefaultWebhookTolerance)
	header := sign(body, secret, atEdge.Unix())
	if !VerifyWebhookSignature(body, header, secret, VerifyWebhookOptions{Now: now}) {
		t.Fatal("expected a signature exactly at the tolerance edge to pass")
	}
}

func TestVerifyWebhookSignature_OneSecondBeyondEdge(t *testing.T) {
	t.Parallel()
	body := []byte("edge")
	secret := "whsec_xx"
	now := time.Now().Truncate(time.Second)
	beyond := now.Add(-DefaultWebhookTolerance - time.Second)
	header := sign(body, secret, beyond.Unix())
	if VerifyWebhookSignature(body, header, secret, VerifyWebhookOptions{Now: now}) {
		t.Fatal("expected a signature one second past the edge to fail")
	}
}

func TestVerifyWebhookSignature_OneSecondBeyondEdgeInTheFuture(t *testing.T) {
	t.Parallel()
	body := []byte("edge")
	secret := "whsec_xx"
	now := time.Now().Truncate(time.Second)
	beyond := now.Add(DefaultWebhookTolerance + time.Second)
	header := sign(body, secret, beyond.Unix())
	if VerifyWebhookSignature(body, header, secret, VerifyWebhookOptions{Now: now}) {
		t.Fatal("expected a future signature one second past the edge to fail")
	}
}

func TestVerifyWebhookSignature_FieldOrderIndependent(t *testing.T) {
	t.Parallel()
	body := []byte("x")
	secret := "whsec_xx"
	now := time.Now()
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(strconv.FormatInt(now.Unix(), 10)))
	mac.Write([]byte("."))
	mac.Write(body)
	sig := hex.EncodeToString(mac.Sum(nil))

	forward := "t=" + strconv.FormatInt(now.Unix(), 10) + ",v1=" + sig
	reverse := "v1=" + sig + ",t=" + strconv.FormatInt(now.Unix(), 10)
	if !VerifyWebhookSignature(body, forward, secret, VerifyWebhookOptions{Now: now}) {
		t.Error("forward header order failed")
	}
	if !VerifyWebhookSignature(body, reverse, secret, VerifyWebhookOptions{Now: now}) {
		t.Error("reverse header order failed")
	}
}

func TestVerifyWebhookSignature_MalformedHeaders(t *testing.T) {
	t.Parallel()
	for _, h := range []string{
		"",
		"garbage",
		"t=abc,v1=zzz",
		"v1=onlysig",
		"t=123",
	} {
		if VerifyWebhookSignature([]byte("x"), h, "whsec_xx", VerifyWebhookOptions{Now: time.Unix(0, 0)}) {
			t.Errorf("expected malformed header %q to fail", h)
		}
	}
}

func TestVerifyWebhookSignature_CustomTolerance(t *testing.T) {
	t.Parallel()
	body := []byte("x")
	secret := "whsec_xx"
	now := time.Now()
	tenMinAgo := now.Add(-10 * time.Minute)
	header := sign(body, secret, tenMinAgo.Unix())
	// 15-minute tolerance accepts a 10-minute-old signature.
	if !VerifyWebhookSignature(body, header, secret, VerifyWebhookOptions{
		Now:       now,
		Tolerance: 15 * time.Minute,
	}) {
		t.Fatal("expected 10min-old signature to pass with 15min tolerance")
	}
}

// V-359 — rotation-grace: HeaderPrev allows acceptance during the 24h
// dual-sign window when the customer hasn't yet rolled the new secret
// across their verifier.
func TestVerifyWebhookSignature_V359_AcceptsPrevWhenOldSecret(t *testing.T) {
	t.Parallel()
	body := []byte(`{"event":"x"}`)
	now := time.Now()
	ts := now.Unix()
	newSecret := "whsec_new_rotated"
	oldSecret := "whsec_old_pre_rotation"
	header := sign(body, newSecret, ts)
	headerPrev := sign(body, oldSecret, ts)

	// Customer hasn't rolled forward — verifier still uses oldSecret.
	if !VerifyWebhookSignature(body, header, oldSecret, VerifyWebhookOptions{
		Now:        now,
		HeaderPrev: headerPrev,
	}) {
		t.Fatal("expected oldSecret + headerPrev to pass during grace")
	}
}

func TestVerifyWebhookSignature_V359_AcceptsCurrentWhenNewSecret(t *testing.T) {
	t.Parallel()
	body := []byte(`{"event":"x"}`)
	now := time.Now()
	ts := now.Unix()
	newSecret := "whsec_new_rotated"
	oldSecret := "whsec_old_pre_rotation"
	header := sign(body, newSecret, ts)
	headerPrev := sign(body, oldSecret, ts)

	// Customer rolled forward — verifier uses newSecret.
	if !VerifyWebhookSignature(body, header, newSecret, VerifyWebhookOptions{
		Now:        now,
		HeaderPrev: headerPrev,
	}) {
		t.Fatal("expected newSecret + current header to pass during grace")
	}
}

func TestVerifyWebhookSignature_V359_RejectsWhenNeitherMatches(t *testing.T) {
	t.Parallel()
	body := []byte("x")
	now := time.Now()
	ts := now.Unix()
	header := sign(body, "whsec_new_rotated", ts)
	headerPrev := sign(body, "whsec_old_pre_rotation", ts)

	if VerifyWebhookSignature(body, header, "whsec_unrelated", VerifyWebhookOptions{
		Now:        now,
		HeaderPrev: headerPrev,
	}) {
		t.Fatal("expected unrelated secret to reject both headers")
	}
}

func TestVerifyWebhookSignature_V359_HeaderPrevEmptyKeepsSingleHeaderBehavior(t *testing.T) {
	t.Parallel()
	body := []byte("x")
	now := time.Now()
	ts := now.Unix()
	secret := "whsec_xx"
	header := sign(body, secret, ts)

	if !VerifyWebhookSignature(body, header, secret, VerifyWebhookOptions{Now: now}) {
		t.Fatal("expected single-header verify to still work")
	}
}

// Hex is case-insensitive, so the verifier must be too. The comparison decodes
// both sides before checking, which is what makes casing irrelevant.
//
// Guarded here because it was not: replacing the decode with a hex-TEXT compare
// leaves every other arm in this file green. sdk-python carried exactly that
// shape and was the only one of the three to refuse an upper-case signature for
// the same body, secret and timestamp — a webhook one customer verified was one
// their neighbour could not.
//
// Nothing is weakened: the HMAC still has to match byte for byte via hmac.Equal.
// This API signs lowercase, so other casings arrive from a proxy or a
// hand-built request.
func TestVerifyWebhookSignature_UpperCaseHexAccepted(t *testing.T) {
	body := []byte(`{"event":"x"}`)
	secret := "whsec_case"
	ts := time.Now().Unix()
	header := "t=" + strconv.FormatInt(ts, 10) + ",v1=" + strings.ToUpper(sigHex(body, secret, ts))
	if !VerifyWebhookSignature(body, header, secret) {
		t.Fatal("an upper-case signature was refused; hex encodes the same bytes either way")
	}
}

func TestVerifyWebhookSignature_MixedCaseHexAccepted(t *testing.T) {
	body := []byte(`{"event":"x"}`)
	secret := "whsec_case"
	ts := time.Now().Unix()
	h := sigHex(body, secret, ts)
	mixed := []rune(h)
	for i := range mixed {
		if i%2 == 0 {
			mixed[i] = []rune(strings.ToUpper(string(mixed[i])))[0]
		}
	}
	header := "t=" + strconv.FormatInt(ts, 10) + ",v1=" + string(mixed)
	if !VerifyWebhookSignature(body, header, secret) {
		t.Fatal("a mixed-case signature was refused; this is not a second lookup table")
	}
}

// Without this, accepting everything would satisfy the two arms above.
func TestVerifyWebhookSignature_CaseInsensitivityDoesNotWeaken(t *testing.T) {
	body := []byte(`{"event":"x"}`)
	secret := "whsec_case"
	ts := time.Now().Unix()
	tsStr := strconv.FormatInt(ts, 10)

	wrong := "t=" + tsStr + ",v1=" + strings.ToUpper(sigHex(body, "whsec_other", ts))
	if VerifyWebhookSignature(body, wrong, secret) {
		t.Fatal("an upper-cased signature from the WRONG secret verified")
	}
	if VerifyWebhookSignature(body, "t="+tsStr+",v1="+strings.Repeat("z", 64), secret) {
		t.Fatal("a non-hex signature verified")
	}
	odd := sigHex(body, secret, ts)
	if VerifyWebhookSignature(body, "t="+tsStr+",v1="+odd[:len(odd)-1], secret) {
		t.Fatal("an odd-length hex signature verified")
	}
}

// V-2010 — hmac.New accepts a zero-length key and returns a perfectly good
// digest, so an attacker who knows the body and timestamp could compute
// HMAC-SHA256("", "<t>.<body>") and VerifyWebhookSignature returned true. The
// forgery below is built with the EMPTY key on purpose: a signature made with a
// real secret would be refused for the ordinary reason and prove nothing about
// this branch. The doc comment promises "returns false on any failure mode".
func TestVerifyWebhookSignature_EmptySecretRefusesEmptyKeyForgery(t *testing.T) {
	body := []byte(`{"id":"evt_1"}`)
	ts := time.Now().Unix()
	tsStr := strconv.FormatInt(ts, 10)
	mac := hmac.New(sha256.New, []byte(""))
	mac.Write([]byte(tsStr + "." + string(body)))
	forged := hex.EncodeToString(mac.Sum(nil))
	if VerifyWebhookSignature(body, "t="+tsStr+",v1="+forged, "") {
		t.Fatal("an empty secret verified an HMAC forged with the empty key")
	}
}

File: packages/sdk-typescript/tests/unit/webhook-signature.test.ts
import { createHmac } from 'node:crypto';
import { afterEach, describe, expect, it, vi } from 'vitest';
import { verifyWebhookSignature } from '../../src/webhook-signature.js';

const SECRET = 'whsec_test_supersecret';

function sign(body: string, timestamp: number, secret = SECRET): string {
  const hex = createHmac('sha256', secret).update(`${timestamp.toString()}.${body}`).digest('hex');
  return `t=${timestamp.toString()},v1=${hex}`;
}

describe('verifyWebhookSignature', () => {
  it('accepts a valid signature with current timestamp', async () => {
    const now = Date.now();
    const t = Math.floor(now / 1000);
    const body = '{"event":"session.completed"}';
    const ok = await verifyWebhookSignature({
      body,
      header: sign(body, t),
      secret: SECRET,
      nowMs: now,
    });
    expect(ok).toBe(true);
  });

  it('rejects when secret differs', async () => {
    const now = Date.now();
    const t = Math.floor(now / 1000);
    const body = 'x';
    const ok = await verifyWebhookSignature({
      body,
      header: sign(body, t, 'wrong-secret'),
      secret: SECRET,
      nowMs: now,
    });
    expect(ok).toBe(false);
  });

  it('rejects when body is tampered', async () => {
    const now = Date.now();
    const t = Math.floor(now / 1000);
    const ok = await verifyWebhookSignature({
      body: 'tampered',
      header: sign('original', t),
      secret: SECRET,
      nowMs: now,
    });
    expect(ok).toBe(false);
  });

  it('rejects timestamps outside tolerance window', async () => {
    const now = Date.now();
    const t = Math.floor(now / 1000) - 600; // 10 minutes old
    const body = 'x';
    const ok = await verifyWebhookSignature({
      body,
      header: sign(body, t),
      secret: SECRET,
      nowMs: now,
    });
    expect(ok).toBe(false);
  });

  it('accepts timestamps within configured tolerance', async () => {
    const now = Date.now();
    const t = Math.floor(now / 1000) - 200; // ~3 minutes old (within default 5 min)
    const body = 'x';
    const ok = await verifyWebhookSignature({
      body,
      header: sign(body, t),
      secret: SECRET,
      nowMs: now,
    });
    expect(ok).toBe(true);
  });

  // ── the FUTURE half of the skew window ──────────────────────────────────────
  //
  // The two arms above walk the PAST only. `sdk-typescript-webhook-signature-content-parity`
  // states why that is not enough, in its own words: "the `Math.abs` is load-bearing —
  // it catches BOTH past-stale (replay) AND future-clock-skew (attacker forging a
  // timestamp far in the future). Drift to one-sided `now - timestampMs >` would let
  // future-dated signatures slip through." It asserts the SUBSTRING `Math.abs`, which
  // any rewrite that keeps the call while changing the comparison would satisfy.
  //
  // MEASURED: replacing `Math.abs(now - parsed.timestampMs)` with
  // `(now - parsed.timestampMs)` — one-sided, so a future-dated signature never
  // expires — left all 244 SDK unit tests green.
  //
  // ⚠️ The other two SDKs already cover this. sdk-python has
  // `test_rejects_future_timestamp_outside_tolerance` and sdk-go has `farFuture :=
  // now.Add(10 * time.Minute)`; TypeScript was the odd one out on a security property
  // its own siblings check. Cross-SDK parity is not a nicety here — a customer picks
  // one SDK and gets whatever that one enforces.
  it('rejects a FUTURE-dated timestamp outside tolerance. A forged far-future timestamp is a signature that never expires: it stays inside any past-only window forever, so the replay bound the header exists to impose simply does not apply to it.', async () => {
    const now = Date.now();
    const t = Math.floor(now / 1000) + 600; // 10 minutes ahead
    const body = 'x';
    const ok = await verifyWebhookSignature({
      body,
      header: sign(body, t),
      secret: SECRET,
      nowMs: now,
    });
    expect(ok, 'a signature dated 10 minutes in the future was accepted').toBe(false);
  });

  // The comparison is `> tolerance * 1000` — exclusive — so exactly-at-tolerance is
  // INSIDE the window. Both edges are driven because an off-by-one that flipped the
  // operator would still pass every arm above: 200s and 600s are nowhere near 300s.
  it('accepts a timestamp exactly AT the tolerance edge, past side. The check is exclusive (`>`), so the boundary itself is inside the window — pinning it means a flip to `>=` is a failure rather than a silent one-second narrowing.', async () => {
    const now = Date.now();
    const t = Math.floor(now / 1000) - 300; // default tolerance is 300s
    const body = 'edge';
    const ok = await verifyWebhookSignature({
      body,
      header: sign(body, t),
      secret: SECRET,
      nowMs: Math.floor(now / 1000) * 1000,
    });
    expect(ok, 'a signature exactly at the tolerance edge was rejected').toBe(true);
  });

  it('rejects one second BEYOND the tolerance edge, past side. With the arm above this brackets the boundary, so a widened window cannot pass as "still within tolerance".', async () => {
    const now = Date.now();
    const t = Math.floor(now / 1000) - 301;
    const body = 'edge';
    const ok = await verifyWebhookSignature({
      body,
      header: sign(body, t),
      secret: SECRET,
      nowMs: Math.floor(now / 1000) * 1000,
    });
    expect(ok, 'a signature one second past the edge was accepted').toBe(false);
  });

  it('rejects one second beyond the tolerance edge, FUTURE side. The same bracket on the half nothing was walking — and the half a one-sided comparison leaves wide open.', async () => {
    const now = Date.now();
    const t = Math.floor(now / 1000) + 301;
    const body = 'edge';
    const ok = await verifyWebhookSignature({
      body,
      header: sign(body, t),
      secret: SECRET,
      nowMs: Math.floor(now / 1000) * 1000,
    });
    expect(ok, 'a signature one second into the future beyond tolerance was accepted').toBe(false);
  });

  it('rejects malformed header', async () => {
    expect(
      await verifyWebhookSignature({ body: 'x', header: 'not-a-valid-header', secret: SECRET }),
    ).toBe(false);
    expect(await verifyWebhookSignature({ body: 'x', header: 't=12345', secret: SECRET })).toBe(
      false,
    );
    expect(await verifyWebhookSignature({ body: 'x', header: 'v1=abc', secret: SECRET })).toBe(
      false,
    );
    expect(await verifyWebhookSignature({ body: 'x', header: undefined, secret: SECRET })).toBe(
      false,
    );
  });

  it('accepts Uint8Array body', async () => {
    const now = Date.now();
    const t = Math.floor(now / 1000);
    const bodyText = '{"x":1}';
    const body = new TextEncoder().encode(bodyText);
    const ok = await verifyWebhookSignature({
      body,
      header: sign(bodyText, t),
      secret: SECRET,
      nowMs: now,
    });
    expect(ok).toBe(true);
  });

  // Dual-`v1=` in ONE header (Stripe-style multi-signature). During a
  // secret-rotation grace window the server dual-signs into a single
  // `x-driftstack-signature: t=,v1=<new>,v1=<old>` header (no separate
  // -prev header). The verifier must accept if EITHER signature matches the
  // configured secret — regardless of order — so a customer on the new OR
  // the old secret verifies during grace.
  describe('multi-v1 single header (rotation dual-sign)', () => {
    const NEW = 'whsec_new_rotated';
    const OLD = 'whsec_old_pre_rotation';
    const h = (b: string, t: number, s: string) =>
      createHmac('sha256', s).update(`${t.toString()}.${b}`).digest('hex');

    it('accepts the new-secret holder when v1=<new> is FIRST (was the discarded position)', async () => {
      const now = Date.now();
      const t = Math.floor(now / 1000);
      const body = '{"event":"x"}';
      const header = `t=${t.toString()},v1=${h(body, t, NEW)},v1=${h(body, t, OLD)}`;
      expect(await verifyWebhookSignature({ body, header, secret: NEW, nowMs: now })).toBe(true);
    });

    it('accepts the old-secret holder when v1=<old> is LAST', async () => {
      const now = Date.now();
      const t = Math.floor(now / 1000);
      const body = '{"event":"x"}';
      const header = `t=${t.toString()},v1=${h(body, t, NEW)},v1=${h(body, t, OLD)}`;
      expect(await verifyWebhookSignature({ body, header, secret: OLD, nowMs: now })).toBe(true);
    });

    it('rejects a secret matching NEITHER signature', async () => {
      const now = Date.now();
      const t = Math.floor(now / 1000);
      const body = '{"event":"x"}';
      const header = `t=${t.toString()},v1=${h(body, t, NEW)},v1=${h(body, t, OLD)}`;
      expect(
        await verifyWebhookSignature({ body, header, secret: 'whsec_unrelated', nowMs: now }),
      ).toBe(false);
    });
  });

  // V-359 — rotation grace: when the customer hasn't yet rolled the new
  // secret across their verifier, they pass `headerPrev` (the prev
  // signature header from the inbound request); the verifier accepts
  // EITHER header matching the supplied secret.
  describe('rotation grace (headerPrev)', () => {
    const NEW_SECRET = 'whsec_new_rotated';
    const OLD_SECRET = 'whsec_old_pre_rotation';

    it('accepts when only the prev header matches the (old) secret', async () => {
      const now = Date.now();
      const t = Math.floor(now / 1000);
      const body = '{"event":"x"}';
      // Server signs with both secrets; customer hasn't yet rolled
      // forward, so they verify against OLD_SECRET. Prev header
      // contains the OLD-secret HMAC.
      const ok = await verifyWebhookSignature({
        body,
        header: sign(body, t, NEW_SECRET),
        headerPrev: sign(body, t, OLD_SECRET),
        secret: OLD_SECRET,
        nowMs: now,
      });
      expect(ok).toBe(true);
    });

    it('accepts when only the current header matches the (new) secret', async () => {
      const now = Date.now();
      const t = Math.floor(now / 1000);
      const body = '{"event":"x"}';
      // Customer has rolled forward to NEW_SECRET.
      const ok = await verifyWebhookSignature({
        body,
        header: sign(body, t, NEW_SECRET),
        headerPrev: sign(body, t, OLD_SECRET),
        secret: NEW_SECRET,
        nowMs: now,
      });
      expect(ok).toBe(true);
    });

    it('rejects when neither header matches the configured secret', async () => {
      const now = Date.now();
      const t = Math.floor(now / 1000);
      const body = 'x';
      const ok = await verifyWebhookSignature({
        body,
        header: sign(body, t, NEW_SECRET),
        headerPrev: sign(body, t, OLD_SECRET),
        secret: 'whsec_unrelated',
        nowMs: now,
      });
      expect(ok).toBe(false);
    });

    it('headerPrev undefined keeps single-header behavior', async () => {
      const now = Date.now();
      const t = Math.floor(now / 1000);
      const body = 'x';
      const ok = await verifyWebhookSignature({
        body,
        header: sign(body, t, NEW_SECRET),
        secret: NEW_SECRET,
        nowMs: now,
      });
      expect(ok).toBe(true);
    });
  });
});

// V-1408 — two paths in the SDK's verifier that nothing exercised. This is code the
// CUSTOMER runs, in their own runtime, to decide whether a webhook is genuine, so both
// failure directions are theirs to live with: rejecting a real delivery, or accepting a
// forged one.
describe('verifyWebhookSignature — the header may arrive as an array, and crypto may be absent', () => {
  afterEach(() => {
    vi.unstubAllGlobals();
  });

  it('CRITICAL accepts a header delivered as a string ARRAY, which is how a duplicated `driftstack-signature` reaches a Node handler. The input type has always said `string | string[]`; nothing had ever passed the array form, and treating it as a plain string makes it fail the `typeof` check and reject a genuine delivery — a customer-visible outage with a correct signature on the wire.', async () => {
    const now = Date.now();
    const t = Math.floor(now / 1000);
    const body = '{"event":"session.completed"}';
    const header = sign(body, t);

    await expect(
      verifyWebhookSignature({ body, header: [header], secret: SECRET, nowMs: now }),
    ).resolves.toBe(true);
    await expect(
      verifyWebhookSignature({ body, header: [header, header], secret: SECRET, nowMs: now }),
    ).resolves.toBe(true);
  });

  it('CRITICAL reads the FIRST value of the array. A proxy that duplicates the header repeats the genuine one; anything appended after it is the part an attacker could influence, so first-wins is the safe reading and the one that must not drift to last-wins.', async () => {
    const now = Date.now();
    const t = Math.floor(now / 1000);
    const body = '{"event":"session.completed"}';
    const genuine = sign(body, t);
    const forged = sign(body, t, 'whsec_attacker_secret');

    await expect(
      verifyWebhookSignature({ body, header: [genuine, forged], secret: SECRET, nowMs: now }),
    ).resolves.toBe(true);
    await expect(
      verifyWebhookSignature({ body, header: [forged, genuine], secret: SECRET, nowMs: now }),
      'a forged value first must not be rescued by a genuine one after it',
    ).resolves.toBe(false);
  });

  it('CRITICAL fails CLOSED when the runtime has no WebCrypto. The SDK ships into browsers and older Node, and the probe for `crypto.subtle` is deliberate — but an absent primitive must mean "cannot verify, reject", never a throw the customer catches into an accept-by-default, and never a silent true.', async () => {
    const now = Date.now();
    const t = Math.floor(now / 1000);
    const body = '{"event":"session.completed"}';
    const header = sign(body, t);

    vi.stubGlobal('crypto', undefined);
    await expect(
      verifyWebhookSignature({ body, header, secret: SECRET, nowMs: now }),
    ).resolves.toBe(false);

    vi.stubGlobal('crypto', {});
    await expect(
      verifyWebhookSignature({ body, header, secret: SECRET, nowMs: now }),
      'a crypto object without subtle is the same situation and must answer the same way',
    ).resolves.toBe(false);
  });
  // V-2010 — an empty secret is the one input where all three SDKs answered
  // differently and all three were wrong: Python and Go hashed with a zero-length
  // key and VERIFIED an HMAC an attacker computes with no secret at all, and this
  // one threw DataError out of subtle.importKey where the contract promises a
  // boolean. The forged header below is built with the empty key on purpose — a
  // signature made with SECRET would be refused for the ordinary reason and prove
  // nothing about this branch.
  it('CRITICAL an empty secret refuses a signature forged with the empty key, and returns false rather than throwing', async () => {
    const body = JSON.stringify({ id: 'evt_1' });
    const t = Math.floor(Date.now() / 1000);
    const forged = createHmac('sha256', '').update(`${t}.${body}`).digest('hex');
    await expect(
      verifyWebhookSignature({ body, header: `t=${t},v1=${forged}`, secret: '' }),
      'an attacker who knows the body and timestamp must not verify against an empty secret',
    ).resolves.toBe(false);
  });
});
